clusterloader2/pkg/util/cluster.go

```diff
@@ -0,0 +1,185 @@
+	schedulercache "k8s.io/kubernetes/pkg/scheduler/cache"
+)
+
+var (
+	// UnreachableTaintTemplate is the taint for when a node becomes unreachable.
+	unreachableTaintTemplate = &corev1.Taint{
+		Key:    algorithm.TaintNodeUnreachable,
+		Effect: corev1.TaintEffectNoExecute,
+	}
+
+	// NotReadyTaintTemplate is the taint for when a node is not ready for
+	// executing pods
+	notReadyTaintTemplate = &corev1.Taint{
+		Key:    algorithm.TaintNodeNotReady,
+		Effect: corev1.TaintEffectNoExecute,
+	}
+)
+
+// GetNodesNumber returns number of nodes in the cluster.
+func GetNodesNumber(c clientset.Interface) (int, error) {
+	nodeList, err := c.CoreV1().Nodes().List(metav1.ListOptions{})
+	if err != nil {
+		return 0, err
+	}
+	numNodes := 0
+	for i := range nodeList.Items {
+		if isNodeSchedulable(&nodeList.Items[i]) && isNodeUntainted(&nodeList.Items[i]) {
+			numNodes++
+		}
+	}
+	return numNodes, err
+}
+
+// Node is schedulable if:
+// 1) doesn't have "unschedulable" field set
+// 2) it's Ready condition is set to true
+// 3) doesn't have NetworkUnavailable condition set to true
+func isNodeSchedulable(node *corev1.Node) bool {
+	nodeReady := isNodeConditionSetAsExpected(node, corev1.NodeReady, true, false)
+	networkReady := isNodeConditionUnset(node, corev1.NodeNetworkUnavailable) ||
+		isNodeConditionSetAsExpected(node, corev1.NodeNetworkUnavailable, false, true)
+	return !node.Spec.Unschedulable && nodeReady && networkReady
+}
+
+// Test whether a fake pod can be scheduled on "node", given its current taints.
+func isNodeUntainted(node *corev1.Node) bool {
+	fakePod := &corev1.Pod{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "Pod",
+			APIVersion: "v1",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "fake-not-scheduled",
+			Namespace: "fake-not-scheduled",
+		},
+		Spec: corev1.PodSpec{
+			Containers: []corev1.Container{
+				{
+					Name:  "fake-not-scheduled",
+					Image: "fake-not-scheduled",
+				},
+			},
+		},
+	}
+	nodeInfo := schedulercache.NewNodeInfo()
+	nodeInfo.SetNode(node)
+	fit, err := podToleratesNodeTaints(fakePod, nodeInfo, func(t *corev1.Taint) bool {
+		// PodToleratesNodeTaints is only interested in NoSchedule and NoExecute taints.
+		return t.Effect == corev1.TaintEffectNoSchedule || t.Effect == corev1.TaintEffectNoExecute
+	})
+	//fit, _, err := predicates.PodToleratesNodeTaints(fakePod, nil, nodeInfo)
+	if err != nil {
+		glog.Errorf("Can't test predicates for node %s: %v", node.Name, err)
+		return false
+	}
+	return fit
+}
+
+func podToleratesNodeTaints(pod *corev1.Pod, nodeInfo *schedulercache.NodeInfo, filter func(t *corev1.Taint) bool) (bool, error) {
+	taints, err := nodeInfo.Taints()
+	if err != nil {
+		return false, err
+	}
+
+	if v1helper.TolerationsTolerateTaintsWithFilter(pod.Spec.Tolerations, taints, filter) {
+		return true, nil
+	}
+	return false, nil
+}
+
+func isNodeConditionSetAsExpected(node *corev1.Node, conditionType corev1.NodeConditionType, wantTrue, silent bool) bool {
```

Comment:
Copying this function is unfortunate - it's pretty complicated...

Reply:
The problem is that original function is located at `kubernete/test/e2e/framework`. If I vendor this package, I also need to vendor the whole kubernetes.
I will try to simplify to the code.